### DISCUSSION TITLE
lockFileMaintenance vs rangeStrategy update-lockfile

### DISCUSSION THREAD

**Discussion by ryancausey** [renovatebot/renovate#19285] [Request Help] [answered]
It seems like there's two separate methods for updating and maintaining lock files. There's setting the [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy) to `update-lockfile` and then there's [lockFileMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance). Are these two things mutually exclusive or can they be used together? If they can be used together, how do they interact? Does one take precedence over the other?

**Comment by rarkins** [comment, marked as answer]
lock file maintenance is something you'd typically do periodically such as weekly to do a full refresh of the lock file, including direct dependencies and all transitive. It's essentially "delete the lock file and recalculate it".

rangeStrategy=update-lockfile relates to direct dependencies only, where you have a range in your package file but wish Renovate to update the locked version whenever a new release is available. It has no direct effect on transitive dependencies

**Reply by ryancausey** [reply]
Just adding some behavior I've seen while testing out a configuration with both a [rangeStrategy](https://docs.renovatebot.com/configuration-options/#rangestrategy) of `update-lockfile` and [lockFIleMaintenance](https://docs.renovatebot.com/configuration-options/#lockfilemaintenance) enabled for future travelers:

If there are in-range updates, both a lock file maintenance PR and individual PRs for the individual dependencies will be created, potentially duplicating effort. I determined this by merging the lock file maintenance PR and watching renovate close the other PRs on subsequent runs as lock file maintenance had already bumped the dependencies in the individual PRs opened by the `update-lockfile` range strategy.

It seems that they are two separate logic paths within renovate and do not consider one another, so these two configuration values overlap on the lock file updates portion. This makes me think that the better configuration to combine the two, assuming one does not want to pin versions, is:

```json
{
    "rangeStrategy": "replace",
    "lockFileMaintenance": {
        "enabled": true
    }
}
```

My understanding is that this way, there won't be PRs opened by the range strategy for dependencies that will be updated via lock file maintenance. Also, any time a new update falls outside the defined semver range, the range in the package manager file will be updated by the range strategy of `replace`.

The only thing I might suggest is some notes in the documentation regarding how these two behaviors interact since it took me some time and experimentation to come to the conclusion I have here, but otherwise it seems to work great once configured right.
